p3/plot/backend/pgfplots.py

```diff
@@ -0,0 +1,434 @@
+        for app, mark in zip(applications, markers):
+            app_line_specs[
+                app
+            ] = f"{app_to_tex_name[app]}, thick, solid, mark={mark}"
+
+        # Choose labels for each platform
+        plat_labels = dict(zip(platforms, string.ascii_uppercase))
+
+        # Set the number of rows in the platform key (if set)
+        # NOTE: This is different to matplotlib because PGF plots uses columns,
+        #       and then transposes (so columns become rows)
+        plat_legend_nrows = plat_legend.kwargs["nrows"]
+
+        plotylabel = ""
+        if eff_column == "app eff":
+            plotylabel = "Application Efficiency"
+        elif eff_column == "arch eff":
+            plotylabel = "Architectural Efficiency"
+        else:
+            msg = "Unexpected efficiency column name: %s"
+            raise ValueError(msg % (eff_column))
+
+        # Build an dictionary with "application name, [ data ]" for the
+        # efficiency plot
+        plot_effs = {}
+        for app_name in applications:
+            app_df = df[
+                (df["application"] == app_name) & (df[eff_column] > 0.0)
+            ]
+            app_df = app_df.sort_values(by=[eff_column], ascending=False)
+            yvalues = app_df[eff_column]
+            supported_platforms = list(app_df["platform"])
+            xvalues = np.arange(1, len(supported_platforms) + 1)
+            plot_effs[app_name] = tuple(zip(xvalues, yvalues))
+
+        # Build a dictionary with "application name, (application name, qp)"
+        # for the pp bar plot
+        pp = p3.metrics.pp(df)
+        pp_column = eff_column.replace("eff", "pp")
+        pp_bars = {
+            app: f"({app}, {app_pp})"
+            for app, app_pp in zip(pp["application"], pp[pp_column])
+        }
+
+        # Build a dictionary with "application name, [ A, B, C, etc ]", etc
+        # for the platform plot
+        plat_plot = {}
+        for i, app_name in enumerate(applications):
+            app_df = df[
+                (df["application"] == app_name) & (df[eff_column] > 0.0)
+            ]
+            app_df = app_df.sort_values(by=[eff_column], ascending=False)
+            supported_platforms = list(app_df["platform"])
+            supported_alpha = [plat_labels[s] for s in supported_platforms]
+            plat_plot[app_name] = supported_alpha
+
+        # Build a jinja environment that can parse the TeX template
+        latex_jinja_env = jinja2.Environment(
+            block_start_string=r"\BLOCK{",
+            block_end_string="}",
+            variable_start_string=r"\VAR{",
+            variable_end_string="}",
+            comment_start_string=r"\#{",
+            comment_end_string="}",
+            line_statement_prefix="%-",
+            line_comment_prefix="%#",
+            trim_blocks=True,
+            autoescape=True,
+            loader=jinja2.FileSystemLoader(
+                os.path.dirname(p3.__file__) + "/plot/backend/textemplates/",
```

Comment:
From what I can tell, this doesn't work.

A simple example of just 
```
p = pkgutil.get_data("p3", "/plot/")
print(p)
```
shows that the `get_data` function wants to load a file, and doesn't like being given a directory name.

Reply:
Is there a reason to avoid the `os` package?

A simple alternative is:
```
p3.__path__[0] + "/plot/backend/templates/"
```

Reply:
Ah, sorry about that.  I forgot that `pkgutil` doesn't work with directories.

We shouldn't use `os` and walk the filesystem where Python provides functionality to do it for us.  I don't understand all the complexities of the Python import system or how packages are guaranteed to be structured after they're installed.  @laserkelvin might have more information about that.

In this case, it seems like jinja actually provides a dedicated function for this purpose.  Can you try:
```
loader = jinja2.PackageLoader("p3.plot.backend")
```

According to the documentation, this will already search for a directory called "templates", so we don't need to specify that.

Reply:
Good spot. That works nicely :)